opensearch-operator/pkg/reconcilers/userrolebinding.go

```diff
@@ -0,0 +1,359 @@
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/go-logr/logr"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/utils/pointer"
+	opsterv1 "opensearch.opster.io/api/v1"
+	"opensearch.opster.io/opensearch-gateway/requests"
+	"opensearch.opster.io/opensearch-gateway/services"
+	"opensearch.opster.io/pkg/helpers"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type UserRoleBindingReconciler struct {
+	client.Client
+	ReconcilerOptions
+	ctx      context.Context
+	osClient *services.OsClusterClient
+	recorder record.EventRecorder
+	instance *opsterv1.OpensearchUserRoleBinding
+	cluster  *opsterv1.OpenSearchCluster
+	logger   logr.Logger
+}
+
+func NewUserRoleBindingReconciler(
+	ctx context.Context,
+	client client.Client,
+	recorder record.EventRecorder,
+	instance *opsterv1.OpensearchUserRoleBinding,
+	opts ...ReconcilerOption,
+) *UserRoleBindingReconciler {
+	options := ReconcilerOptions{}
+	options.apply(opts...)
+	return &UserRoleBindingReconciler{
+		Client:            client,
+		ReconcilerOptions: options,
+		ctx:               ctx,
+		recorder:          recorder,
+		instance:          instance,
+		logger:            log.FromContext(ctx).WithValues("reconciler", "user"),
+	}
+}
+
+func (r *UserRoleBindingReconciler) Reconcile() (retResult ctrl.Result, retErr error) {
+	var reason string
+
+	defer func() {
+		// Skip status updates when option is set
+		if !pointer.BoolDeref(r.updateStatus, true) {
+			return
+		}
+		// When the reconciler is done, figure out what the state of the resource is
+		// is and set it in the state field accordingly.
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(r.instance), r.instance); err != nil {
+				return err
+			}
+			r.instance.Status.Reason = reason
+			if retErr != nil {
+				r.instance.Status.State = opsterv1.OpensearchUserRoleBindingStateError
+			}
+			if retResult.Requeue {
+				r.instance.Status.State = opsterv1.OpensearchUserRoleBindingPending
+			}
+			if retErr == nil && !retResult.Requeue {
+				r.instance.Status.ProvisionedRoles = r.instance.Spec.Roles
+				r.instance.Status.ProvisionedUsers = r.instance.Spec.Users
+				r.instance.Status.State = opsterv1.OpensearchUserRoleBindingStateCreated
+			}
+			return r.Status().Update(r.ctx, r.instance)
+		})
+
+		if err != nil {
+			r.logger.Error(err, "failed to update status")
+		}
+	}()
+
```

Comment:
Not sure if this is possible without making a mess, but the code for each of the reconcilers in the beginning to wait for the cluster and check if its the same seems to be the same, so it might make sense to move that logic to a utility function.

Reply:
It is similar but not entirely the same, because the objects are different.  This may be something we could improve on with generics (once we upgrade the go version), but in the meantime I think this is better.  There is some code duplication, but we have a better abstraction.

Reply:
Yeah, you are right. Without generics it would be hard to make it clean.